neps/nep-0611.md

```diff
@@ -0,0 +1,370 @@
+
+### Gas Key Pending Transaction Constraints
+
+Unlike access key transactions, gas key transactions are not limited in parallelism; they are only limited
+by the amount of gas these transactions consume. Specifically, for a gas key $G$, the total cost of 
+the transactions signed with $G$ in the pending transactions must not exceed the balance of $G$
+(according to the state that the pending transactions are based on).
+
+This constraint should be good enough to cover cases of adding, funding, removing, or withdrawing from gas
+keys as well. For adding and funding, we do need the execution to catch up before the newly available
+balance can be used for pending transactions, which is suboptimal but correct. For withdrawing, the
+action of withdrawing itself goes into the cost of the transaction and therefore is accounted for (and
+note that contract execution cannot create receipts that withdraw from gas keys; only gas key
+transactions can). For deletion, balance in gas keys are not refunded, so although subsequent pending
+transactions may end up failing, the gas committed to those transactions are already burnt, eliminating
+the opportunity for the aforementioned attack.
+
+### Pending Transaction Queue
+
+We would now maintain a new data structure that stores the pending transactions, called the Pending
+Transaction Queue. Although it's conceptually a queue, it is stored as a collection indexed by
+(account ID, transaction type), and further indexed by the block hash. Furthermore, the pending
+transaction queue is stored per shard, not as a single data structure. The contents of the queue
+is exactly the pending transactions according to the definition above.
+
+The constraints are enforced as described above; we reiterate them precisely here. For each account, Let
+$T_A$ be the set of transactions in the queue signed with any access key of this account, and let
+$T_G$
+be the set of transactions in the queue signed with any gas key of this account. The following constraints
+must hold true:
+
+* If the account has a contract deployed (as of the latest available state), then $|T_A| \le P_{\mathrm {max}}$.
+* If any transaction $t\in T_A\cup T_G$ contains a `DeployContract` action, then $|T_A| \setminus \{t\}=\emptyset$. In other words, deploying a contract cannot be done in parallel with any access key
+  transactions.
+* The sum of the costs of all transactions in $T_A$ does not exceed the balance of the account. Here, the cost of a transaction refers to the sum of attached gas and NEAR deposit (as calculated by `total_cost` defined [here](https://github.com/near/nearcore/blob/61c087c5250566070684a624a5ea50e796f61e5f/runtime/runtime/src/config.rs#L332)). Importantly, this cost is knowable by just inspecting the actions contained in the transaction and does not depend on state or execution.
```

Comment:
>  Importantly, this cost is knowable by just inspecting the actions contained in the transaction and does not depend on state or execution.

It does depend on the gas price though, doesn't it? The gas price is part of the block, so not state. I'm not very familiar with SPICE. Is the block header used for execution already known when we do these checks?